src/internal/collection/postgres_collection_test.go

```diff
@@ -64,6 +66,109 @@ func TestPostgresCollections(suite *testing.T) {
+				pgrw := rw.(col.PostgresReadWriteCollection)
+				return pgrw.GetByIndex(TestSecondaryIndex, "foo", &col.TestItem{}, col.DefaultOptions(), func(string) error {
+					return errors.New("GetByIndex callback should not have been called for an empty collection")
+				})
+			})
+			require.NoError(t, err)
+			count, err := emptyRead(context.Background()).Count()
+			require.NoError(t, err)
+			require.Equal(t, int64(0), count)
+		})
+
+		subsuite.Run("Success", func(t *testing.T) {
+			t.Parallel()
+			keys := []string{}
+			err := defaultWriter(context.Background(), func(rw col.ReadWriteCollection) error {
+				testProto := &col.TestItem{}
+				pgrw := rw.(col.PostgresReadWriteCollection)
+				return pgrw.GetByIndex(TestSecondaryIndex, originalValue, testProto, col.DefaultOptions(), func(key string) error {
+					require.Equal(t, testProto.ID, key)
+					require.Equal(t, testProto.Value, originalValue)
+					keys = append(keys, key)
+					// Clear testProto.ID and testProto.Value just to make sure they get overwritten each time
+					testProto.ID = ""
+					testProto.Value = ""
+					return nil
+				})
+			})
+			require.NoError(t, err)
+			require.ElementsEqual(t, keys, idRange(0, defaultCollectionSize))
+			checkDefaultCollection(t, defaultRead, RowDiff{})
+		})
+
+		subsuite.Run("NoResults", func(t *testing.T) {
+			t.Parallel()
+			err := defaultWriter(context.Background(), func(rw col.ReadWriteCollection) error {
+				testProto := &col.TestItem{}
+				pgrw := rw.(col.PostgresReadWriteCollection)
+				return pgrw.GetByIndex(TestSecondaryIndex, changedValue, testProto, col.DefaultOptions(), func(string) error {
+					return errors.New("GetByIndex callback should not have been called for an index value with no rows")
+				})
+			})
+			require.NoError(t, err)
+			checkDefaultCollection(t, defaultRead, RowDiff{})
+		})
+
+		subsuite.Run("InvalidIndex", func(t *testing.T) {
+			t.Parallel()
+			err := defaultWriter(context.Background(), func(rw col.ReadWriteCollection) error {
+				pgrw := rw.(col.PostgresReadWriteCollection)
+				return pgrw.GetByIndex(&col.Index{}, "", &col.TestItem{}, col.DefaultOptions(), func(key string) error {
+					return errors.New("GetByIndex callback should not have been called when using an invalid index")
+				})
+			})
+			require.YesError(t, err)
+			require.Matches(t, "Unknown collection index", err.Error())
+			checkDefaultCollection(t, defaultRead, RowDiff{})
+		})
+
+		subsuite.Run("Nested", func(t *testing.T) {
+			t.Parallel()
+			outerKeys := []string{}
+			innerKeys := []string{}
+			innerID := makeID(3)
+			err := defaultWriter(context.Background(), func(rw col.ReadWriteCollection) error {
+				testProto := &col.TestItem{}
+				pgrw := rw.(col.PostgresReadWriteCollection)
+				return pgrw.GetByIndex(TestSecondaryIndex, originalValue, testProto, col.DefaultOptions(), func(key string) error {
+					outerKeys = append(outerKeys, testProto.ID)
+					if err := pgrw.Get(innerID, testProto); err != nil {
```

Comment:
Maybe it's worth doing a write here, just to enforce what that behavior means.
Probably doesn't matter though